keeper/accounting.go

```diff
@@ -0,0 +1,395 @@
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN "AS IS" BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/types"
+
+	vaultsv2 "dollar.noble.xyz/v3/types/vaults/v2"
+)
+
+// AccountingResult contains the results of a cursor-based accounting operation.
+type AccountingResult struct {
+	PositionsProcessed      uint64
+	TotalPositionsProcessed uint64
+	TotalPositions          uint64
+	Complete                bool
+	NextUser                string
+	AppliedNav              sdkmath.Int
+	YieldDistributed        sdkmath.Int
+}
+
+// updateVaultsV2AccountingWithCursor performs yield accounting with cursor-based pagination.
+// This allows the accounting to be split across multiple message invocations.
+func (k *Keeper) updateVaultsV2AccountingWithCursor(ctx context.Context, maxPositions uint32) (*AccountingResult, error) {
+	navInfo, err := k.GetVaultsV2NAVInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if navInfo.CurrentNav.IsNil() {
+		return nil, fmt.Errorf("current NAV is not set")
+	}
+
+	state, err := k.GetVaultsV2VaultState(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	totalShares, err := k.GetVaultsV2TotalShares(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cursor, err := k.GetVaultsV2AccountingCursor(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine if we need to start a new accounting session
+	needsInit := !cursor.InProgress ||
+		cursor.AccountingNav.IsNil() ||
+		!cursor.AccountingNav.Equal(navInfo.CurrentNav) ||
+		!cursor.AccountingNavTimestamp.Equal(navInfo.LastUpdate)
+
+	if needsInit {
+		// If accounting is in progress with a different NAV, we have a problem
+		// We cannot start new accounting while old accounting is incomplete
+		// as this would leave some users with old NAV and some with new NAV
+		if cursor.InProgress {
+			return nil, fmt.Errorf(
+				"accounting already in progress for NAV %s (started at %s, %d/%d positions processed). "+
+					"Cannot start new accounting for NAV %s until current session completes. "+
+					"Continue calling this message to complete the current session",
+				cursor.AccountingNav.String(),
+				cursor.StartedAt.String(),
+				cursor.PositionsProcessed,
+				cursor.TotalPositions,
+				navInfo.CurrentNav.String(),
+			)
+		}
+
+		// Count total positions
+		totalPositions := uint64(0)
+		if err := k.IterateVaultsV2UserPositions(ctx, func(address types.AccAddress, position vaultsv2.UserPosition) (bool, error) {
+			totalPositions++
+			return false, nil
+		}); err != nil {
```

Comment:
Is iterating every position each time going to make using `maxPositions` per update moot when there are a lot of positions?